Support high-availability setups for the interactive tools proxy

domgz opened the pull request:
This PR closes #16138. At the moment, information exchanges between the [interactive tools proxy](https://github.com/galaxyproject/gx-it-proxy) and Galaxy are done via a SQLite database. This makes it impossible to have the high-availability (redundant) setup for the proxy that we would like to have at [useGalaxy.eu](https://usegalaxy.eu/).

To solve the problem, I opted for a relatively simple solution. Since Galaxy already uses SQLAlchemy, I refactored the existing class `InteractiveToolSqlite` from lib/galaxy/managers/interactivetool.py to use SQLAlchemy instead of [Python's built-in `sqlite3` library](https://docs.python.org/3/library/sqlite3.html). The changes are backwards compatible. Setting `interactivetools_map` to a path makes Galaxy use a SQLite database, just like it did before. Setting it to a [SQLAlchemy database url](https://docs.sqlalchemy.org/en/20/core/engines.html#database-urls) allows to use any database supported by SQLAlchemy (e.g. PostgreSQL). Mappings are written to the table "gxitproxy", just as they did before.

Of course, this also needs support on the [interactive tools proxy](https://github.com/galaxyproject/gx-it-proxy)'s side, which will be contributed separately.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

This is a refactor of the "propagator" `InteractiveToolSqlite` in lib/galaxy/managers/interactivetool.py. Tests are already available and can be run using 
`./run_tests.sh -integration test/integration/test_interactivetools_api.py::TestInteractiveToolsIntegration`. I used standard SQL so that the code is compatible with the widest possible variety of databases.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

domgz committed:
Support using a table in any database supported by SQLAlchemy as mapping for interactive tools

Replace the class `InteractiveToolSqlite` in lib/galaxy/managers/interactivetool.py with a new class `InteractiveToolPropagatorSQLAlchemy`. The new class implements a SQLAlchemy "propagator" for `InteractiveToolManager` (on the same file). This propagator writes the mappings to the table named after the value of `DATABASE_TABLE_NAME`, in the database specified by the SQLAlchemy database url passed to its constructor. Change the constructor of `InteractiveToolManager` so that it uses `InteractiveToolPropagatorSQLAlchemy`.

Change the method `_process_config` of `galaxy.config.GalaxyAppConfiguration` so that it converts the value of `interactivetools_map` to a SQLAlchemy database url if it is a path.

Update documentation to reflect these changes.

domgz commented on lib/galaxy/managers/interactivetool.py:
```diff
@@ -18,125 +19,129 @@
 )
 from galaxy.model.base import transaction
 from galaxy.security.idencoding import IdAsLowercaseAlphanumEncodingHelper
-from galaxy.util.filelock import FileLock
```
I feel like I may be missing something regarding the `FileLock`. What role was it actually playing? Isn't SQLite [already taking care of locks](https://www.sqlite.org/lockingv3.html)?

> The pager module effectively controls access for separate threads, or separate processes, or both. Throughout this document whenever the word "process" is written you may substitute the word "thread" without changing the truth of the statement.

jmchilton replied:
Galaxy has a long history of concurrency issues with respect to SQLite. I cannot promise this FileLock was needed but experience taught me it was better to be safe than sorry here.

domgz replied:
Some information that may help us understand whether the lock is needed in this case.

- [SQLALchemy documentation](https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#database-locking-behavior-concurrency)
> However, no matter what locking modes are used, SQLite will still always lock the database file once a transaction is started and DML (e.g. INSERT, UPDATE, DELETE) has at least been emitted, and this will block other transactions at least at the point that they also attempt to emit DML. By default, the length of time on this block is very short before it times out with an error.
>
> This behavior becomes more critical when used in conjunction with the SQLAlchemy ORM. SQLAlchemy’s [Session](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session) object by default runs within a transaction, and with its autoflush model, may emit DML preceding any SELECT statement. This may lead to a SQLite database that locks more quickly than is expected. The locking mode of SQLite and the pysqlite driver can be manipulated to some degree, however it should be noted that achieving a high degree of write-concurrency with SQLite is a losing battle.

- [SQLite documentation](https://www.sqlite.org/whentouse.html)
> # 2. Situations Where A Client/Server RDBMS May Work Better
>
> - [...]
>
> - **High-volume Websites**
>
>   SQLite will normally work fine as the database backend to a website. But if the website is write-intensive or is so busy that it requires multiple servers, then consider using an enterprise-class client/server database engine instead of SQLite.
>
>
> - [...]
>
> - **High Concurrency**
>   SQLite supports an unlimited number of simultaneous readers, but it will only allow one writer at any instant in time. For many situations, this is not a problem. Writers queue up. Each application does its database work quickly and moves on, and no lock lasts for more than a few dozen milliseconds. But there are some applications that require more concurrency, and those applications may need to seek a different solution. 

From both excerpts I understand that it is likely that the long history of concurrency issues that Galaxy has with SQLite probably has to do with using the SQLAlchemy ORM under high load (maybe you can either support/refute this?).

As far as I know, the interactive tools mapping file is written only by `InteractiveToolSqlite` (or `InteractiveToolPropagatorSQLAlchemy` in this refactor) and read only by Galaxy (the same class) and the proxy. I understand that issues would appear if interactive tools are in extremely high demand (Galaxy constantly writes keys and tokens to the file). However, in that case, wouldn't it just make better sense, as the SQLite docs claim, that the Galaxy admin uses PostgreSQL rather than SQLite? Another solution would be increasing the SQLite lock timeout.

Maybe if @jdavcs agrees to [go forward with a new config property for mappings based on arbitrary SQLAlchemy database URLs](https://github.com/galaxyproject/galaxy/pull/18481#issuecomment-2222244009), it may even make sense to discourage the use of the original property `interactivetools_map` in the Galaxy docs to favour the use of Postgres? After all, [SQLite use is already discouraged](https://docs.galaxyproject.org/en/release_24.1/admin/production.html#switching-to-postgresql-database-server).

In that case though, I think we should also have a look together at the [proxy-side implementation](https://github.com/galaxyproject/gx-it-proxy/pull/21/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R24-R53), because it also has it's drawbacks regarding database setup (maybe it should just be done automatically by Galaxy).

jdavcs replied:
> From both excerpts I understand that it is likely that the long history of concurrency issues that Galaxy has with SQLite probably has to do with using the SQLAlchemy ORM under high load (maybe you can either support/refute this?).

I think the core of the issue here is just SQLite: "SQLite will still always lock the database file once a transaction is started" - it's just suboptimal for write-intensive scenarios. Using the ORM makes it worse: the autoflush model mentioned in the docs ensures that the state of objects loaded from the db is kept in sync with the data in the db by issuing DML statements before retrieving new data (when needed) - so the db file gets locked much more frequently if we rely on the Session. However, we don't use the ORM here.

If we go with [this proposed approached](https://github.com/galaxyproject/galaxy/pull/18481#issuecomment-2222244009), we can still keep SQLite as an option (it may be still a good enough option for smaller instances or dev purposes, etc.; after all we support SQLite as an option for the main database too). We may consider adding a note to [Galaxy's documentation](https://docs.galaxyproject.org/en/master/admin/special_topics/interactivetools.html#interactivetools) and updating the [IT tutorial](https://training.galaxyproject.org/training-material/topics/admin/tutorials/interactive-tools/tutorial.html) accordingly.

domgz replied:
I added a note (dddf5ba) to the documentation and opened galaxyproject/training-material#5178 and galaxyproject/training-material#5179

domgz committed:
Fix bug when using a mapping for interactive tools from a SQLAlchemy database url

domgz committed:
Merge branch 'refs/heads/dev' into interactivetoolssqlalchemy

davelopez reviewed (approved):
Thanks @kysrpex!
LGTM!

domgz committed:
Replace raw SQL queries with SQLAlchemy Core SQL expressions

domgz committed:
Configure SQLAlchemy interactive tool maps via new `interactivetools_map_sqlalchemy` setting

Define a new configuration property `interactivetools_map_sqlalchemy` to store interactivetool proxy mappings on a RDBMS supported by SQLAlchemy.

If set, it takes precedence over `interactivetools_map`. When unset, a SQLite mapping file located at `interactivetools_map` is used (as usual).

domgz committed:
Revert changes to description for setting `sessions` in galaxy.yml.sample

This text is managed by Gravity

domgz committed:
Forbid setting `interactivetools_map_sqlalchemy` from matching `database_connection` or `install_database_connection`

domgz committed:
Add note in interactive tools docs page informing about `interactivetools_map_sqlalchemy` setting

Refer to SQLite docs page describing situations where a client/server RDBMS works better and provide an example. Link to the `interactivetools_map_sqlalchemy` on the Galaxy configuration page.

domgz committed:
workaround `mypy` error on statement using `sqlalchemy.select`

```
lib/galaxy/managers/interactivetool.py:65: error: No overload variant of
"select" matches argument type "ReadOnlyColumnCollection[str, Column[Any]]"
```

domgz commented on lib/galaxy/config/__init__.py:
```diff
@@ -1103,10 +1104,36 @@ def _process_config(self, kwargs: Dict[str, Any]) -> None:
+                    ("install_database_connection", self.install_database_connection),
+                )
+                if value is not None
+            }
```
This is [causing workflow "Test" to fail](https://github.com/galaxyproject/galaxy/actions/runs/10043692668/job/27757024477?pr=18481).

```
____________ ERROR collecting tests/config/test_path_resolves_to.py ____________
ImportError while importing test module '/home/runner/work/galaxy/galaxy/galaxy root/packages/config/tests/config/test_path_resolves_to.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/hostedtoolcache/Python/3.12.4/x64/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/config/test_path_resolves_to.py:5: in <module>
    from galaxy.config import BaseAppConfiguration
galaxy/config/__init__.py:36: in <module>
    from sqlalchemy.engine import make_url as parse_sqlalchemy_url
E   ModuleNotFoundError: No module named 'sqlalchemy'
```

Any hint how to make the module available to the tests?

domgz committed:
Create `gxitproxy` database table when using interactive tools for the first time

domgz committed:
Merge branch 'dev' into interactivetoolssqlalchemy

domgz committed:
Add `sqlalchemy` to config package requirements

`lib.galaxy.config.GalaxyAppConfiguration._process_config` requires `sqlalchemy.engine.make_url` to forbid `interactivetools_map_sqlalchemy` from matching `database_connection` or `install_database_connection` (678bea1)

domgz committed:
Test that Galaxy raises `ConfigurationError` when setting `interactivetools_map_sqlalchemy` matches `database_connection` or `install_database_connection` (678bea1)

jdavcs commented on lib/galaxy/config/__init__.py:
```diff
@@ -1103,10 +1104,36 @@ def _process_config(self, kwargs: Dict[str, Any]) -> None:
-        # InteractiveTools propagator mapping file
-        self.interactivetools_map = self._in_root_dir(
+        # InteractiveTools propagator mapping
+        self.interactivetools_map = "sqlite:///" + self._in_root_dir(
```
Given that `interactivetools_map` is ignored if `interactivetools_map_sqlalchemy` is set, I think we should *first* check `interactivetools_map_sqlalchemy`. If it's not set, then build the sqlite url; but if it is set, then set `interactivetools_map` to None for clarity (and to avoid potential bugs in the future)

domgz replied:
Addressed in 26fd210c

jdavcs commented on lib/galaxy/managers/interactivetool.py:
```diff
@@ -18,125 +26,97 @@
+        columns = ("token", "host", "port", "info")
+
+        with self._engine.connect() as conn:
+            query = select(*gxitproxy.c[columns]).where(
```
Minor thing: would you mind renaming `query`? [Here's why](https://docs.google.com/presentation/d/1Vt_jWo4oM9X7D2uPnYed-dh3rInOK-LI3g4FMRZNYTA/edit#slide=id.g2121adfd7d3_0_183) (we still have a lot of `query` variables that are objects of type `Query`; confusing the two is easy and may lead to very subtle bugs)

domgz replied:
It makes sense, addressed in e040750a

jdavcs commented on lib/galaxy/managers/interactivetool.py:
```diff
@@ -18,125 +26,97 @@
+            )
+            result = conn.execute(query).fetchone()
+
+        return dict(key=key, key_type=key_type, **dict(zip(columns, result))) if result else None
```
This is fine, but I think you could simplify this, and also keep the error handling intact. The columns variable exists only to get rid of minor duplication while making it a bit harder to read; however, we don't even need to enumerate the columns here. Consider this:

```py
stmt = select(proxy).where(proxy.c.key == key, proxy.c.key_type == key_type)
with self._engine.connect() as conn:
    result = conn.exectute(stmt)
    try:
        return dict(zip(result.keys(), result.fetchone()))
    except TypeError:
        log.warning("get(): invalid key: %s key_type %s", key, key_type)
        return None
```
Here, we select all the columns and automatically create the dictionary.

domgz replied:
Damn, so it is not `result.fetchone().keys()`, and the `TypeError` comes from attempting to unpack `None`. Now it makes sense.

Does 6fc0d462 make sense to you?

The `get()` method is dead code by the way (not used anywhere), although I do not see the need to nor benefit of removing it

domgz replied:
mypy is not happy (with either solution I guess), see failed [GitHub actions job](https://github.com/galaxyproject/galaxy/actions/runs/10318754692/job/28565763345?pr=18481)

```
lib/galaxy/managers/interactivetool.py:69: error: Argument 2 to "zip" has
incompatible type "Optional[Row[Any]]"; expected "Iterable[Any]"  [arg-type]
    ..._type=key_type, **dict(zip(cursor_result.keys(), cursor_result.fetchon...
                                                        ^~~~~~~~~~~~~~~~~~~~~~~~ 
```

domgz replied:
I am leaving it like this (see [4e6688c](https://github.com/galaxyproject/galaxy/pull/18481/commits/4e6688cf1113b088017263a982a2f243b88bb64a)) to make mypy happy.

```python
    def get(self, key, key_type):
        with self._engine.connect() as conn:
            stmt = select(*gxitproxy.c["token", "host", "port", "info"]).where(
                gxitproxy.c.key == key,
                gxitproxy.c.key_type == key_type,
            )
            cursor_result = conn.execute(stmt)
            result = cursor_result.fetchone()

            if result is None:
                log.warning("get(): invalid key: %s key_type %s", key, key_type)
                return None

            return dict(key=key, key_type=key_type, **dict(zip(cursor_result.keys(), result)))
```

jdavcs replied:
I've added a comment inline. However, if you're sure this is dead code, then by all means remove it. (unnecessary cognitive load, etc.)

domgz replied:
Dead code removed in 0b6ad177

jdavcs commented on lib/galaxy/managers/interactivetool.py:
```diff
@@ -18,125 +26,97 @@
+
+            # delete existing data with same key
+            query_delete = delete(gxitproxy).where(
+                gxitproxy.c["key"] == key,
```
```suggestion
                gxitproxy.c.key == key,
```
You can access the columns directly in all such cases. The key lookup would be more useful if you needed a subset of the table column collection, like `gxitproxy.c["key", "key_type"]`

domgz replied:
I assume you use this style everywhere so I changed the lookup type in 83050167 for consistency

jdavcs commented on lib/galaxy/config/schemas/config_schema.yml:
```diff
@@ -1520,7 +1520,20 @@ mapping:
         default: interactivetools_map.sqlite
         path_resolves_to: data_dir
         desc: |
```
Can you mention here that the value is ignored if the  `interactivetools_map_sqlalchemy` property is set? (I think we should include this in both property descriptions, not just one)

domgz replied:
Addressed in 018e5417

jdavcs commented on doc/source/admin/galaxy_options.rst:
```diff
@@ -2112,13 +2112,33 @@
 
 :Description:
-    Map for interactivetool proxy.
+    Map for the interactivetool proxy. Mappings are stored on a SQLite
```
```suggestion
    Map for the interactivetool proxy. Mappings are stored in a SQLite
```

domgz replied:
Addressed in ffa8ea4f

jdavcs commented on doc/source/admin/galaxy_options.rst:
```diff
@@ -2112,13 +2112,33 @@
-    Map for interactivetool proxy.
+    Map for the interactivetool proxy. Mappings are stored on a SQLite
+    database file located on this path. As an alternative, you may
+    also store them on any other RDBMS supported by SQLAlchemy using
```
```suggestion
    also store them in any other RDBMS supported by SQLAlchemy using
```

domgz replied:
Addressed in ffa8ea4f

domgz committed:
Set `interactivetools_map` to `None` when `interactivetools_map_sqlalchemy` is defined

Given that `interactivetools_map` is ignored if `interactivetools_map_sqlalchemy` is set, first check `interactivetools_map_sqlalchemy`. If it's not set, then build the sqlite url; but if it is set, then set `interactivetools_map` to `None` for clarity (and to avoid potential bugs in the future)

Co-authored-by: John Davis <jdavcs@gmail.com>

domgz committed:
Fix typos in descriptions of `interactivetools_map` and `interactivetools_map_sqlalchemy`

Authored-by: John Davis <jdavcs@gmail.com>

domgz committed:
Mention that `interactivetools_map_sqlalchemy` overrides `interactivetools_map` in the latter's description

Co-authored-by: John Davis <jdavcs@gmail.com>

domgz committed:
Refactor variable name `query` to `stmt` for calls to `select()`, `insert()` or `delete()`

domgz committed:
Access individual table columns via `__getattr__` rather than `__getitem__`

Match @jdavcs' coding style.

domgz committed:
Restore original error handling for `InteractiveToolPropagatorSQLAlchemy.get()`

Emit a warning when `get()` returns no results.

domgz reviewed (commented):
> @kysrpex thanks, this is looking good! I'll add a few comments inline; however, here's my main question. I've tested this manually and it didn't work correctly for me when enabling the sqlalchemy option: the database table is populated and the job is launched, but when the tool is supposed to become available, I see `Proxy target missing` on a blank screen. It may be due to some obvious local config options I haven't set correctly. However, given that it worked with sqlite settings, I think we should expect it to work with the other setting as well; if any additional config is required, that should be made clear in the documentation. Does it run for you?

@jdavcs Meet the elephant in the room, galaxyproject/gx-it-proxy#21! You need to use that version of the proxy.

> One more general (and minor) comment. Would you be open to using identifiers that are not sqlalchemy-specific? That is, for the config property, the `InteractiveToolPropagatorSQLAlchemy` class and the `parse_sqlalchemy_url` import alias in the config module. Here's why: although we very heavily rely on SQLAlchemy, it's just a tool, and the functionality we use here could be, hypothetically, implemented with another similar tool. So, semantically, it's really database-related, not sqlalchemy-related. The database URL format is, for the most part, an open standard ([ref](https://docs.sqlalchemy.org/en/20/core/engines.html#database-urls)). Finally, from a practical standpoint, it's much easier to manage the code base when grepping for sqlalchemy returns only import statements.
> 
> EDIT: that said, I think it's perfectly OK to mention SQLAlchemy in the config property description, because SQLAlchemy determines what databases are currently supported.

Yes, definitely! I think the same. `interactivetools_map_sqlalchemy` is a terrible choice. Considering that it may even make sense to deprecate `interactivetools_map` in favor of this one at some point, it would be better to choose a name not tied to SQLAlchemy. Something as simple as `interactivetoolsproxymap` might work, but please suggest whatever you have in mind, I think brainstorming it is worth it. Another concern is how long-lived this setting would be. That should have an influence in the name and the decision is on you and the rest of the dev team.

domgz committed:
Workaround to have original error handling for `get()` and pass mypy checks

jdavcs commented on lib/galaxy/managers/interactivetool.py:
```diff
@@ -18,125 +26,100 @@
+                log.warning("get(): invalid key: %s key_type %s", key, key_type)
+                return None
+
+            return dict(key=key, key_type=key_type, **dict(zip(cursor_result.keys(), result)))
```
If you don't select individual columns, i.e. 
`stmt = select(gxitproxy).where(gxitproxy.c.key == key, gxitproxy.c.key_type == key_type)`
then your second return statement can be simplified to 
`return dict(zip(cursor_result.keys(), result))`
Same result.

domgz replied:
Dead code removed, https://github.com/galaxyproject/galaxy/pull/18481#discussion_r1741962053

domgz committed:
Remove dead code `InteractiveToolPropagatorSQLAlchemy.get()`

domgz committed:
Rename `interactivetools_map_sqlalchemy` to `interactivetoolsproxy_map`

domgz committed:
Use `urlparse` to compare `interactivetoolsproxy_map` with `database_connection` and `install_database_connection`

Using `urlparse` rather than `sqlalchemy.engine.make_url` decreases the precision of the comparison, but removes the need to declare `sqlalchemy` as a dependency of the config package.

This is a tradeoff between convenience (preventing mistakes from the user) and complexity (number of dependencies of the config package).

domgz committed:
Revert "Add `sqlalchemy` to config package requirements"

This reverts commit b6ad695d0377f28e0eb492a95d260d1e2be127b7.
